Paternity Test.py

```diff
@@ -0,0 +1,378 @@
+        else:
+            pass
+# For thousands of files this takes a long time, let's try to show some sort of progress.
+# Progress bar packages that use carriage returns don't play nicely with Jupyter when cells
+# are ran more than once, so a balance between the user being updated and not making several
+# lines of text is needed.
+if(df.shape[0] > 100):
+    triggerevery = df.shape[0] // 5
+    nexttrigger = triggerevery
+else:
+    triggerevery = None
+
+logging.info("Child file types are: %s  If this looks wrong, check your GCS directory for missing or junk files." % unique_children)
+
+# Iterate yet again to match children with their parents
+logging.info("Matching children and parents...")
+i = 0 #only used for logging
+j = 0
+for child_extension in unique_children: #loop once per child extension
+    i=i+1
+    logging.info("\t* Looking at %s files (child %d out of %d)" % (child_extension, i, len(unique_children)))
+
+    # Iterate df, where each row represents one file of any type
+    for index_label, row_series in df.iterrows():
+
+        # If user has a big dataframe, this will give a simple progress update
+        if triggerevery is not None:
+            j=j+1
+            if(j==nexttrigger):
+                logger.info("\t\t* %d / %d" % (nexttrigger, df.shape[0]))
+                nexttrigger = nexttrigger + triggerevery
+                if nexttrigger >= df.shape[0]:
+                    nexttrigger = triggerevery
+
+        # If, at this row, we have found a file with the parent extension
+        try:
+            if(df.at[index_label,'FileType'] == PARENT_FILETYPE):
+
+                # Find this parent's child
+                # Child might be above parent so we can't just start from index of the parent
+                for index_label_inception, row_series_inception in df.iterrows():
+                    logging.debug("Outer iter %d, inner iter %d parent %s checking if %s is its child" % (index_label, index_label_inception, df.at[index_label,'filename'], df.at[index_label_inception,'filename']))
+                    if index_label != index_label_inception: #needed to preventing it find itself
+                        if(df.at[index_label,'ID'] == #if parent ID...
+                           df.at[index_label_inception,'ID']) and ( #equals potential child ID, and...
+                            df.at[index_label_inception,'FileType'] == #child's file extension...
+                            child_extension): #equals the current child_extension we are iterating on
+                            # Child found!
+                            logging.debug("    Found "+df.at[index_label_inception,'filename']+" to be child")
+                            df.at[index_label,child_extension] = df.at[index_label_inception,'filename']
+                            df.at[index_label,child_extension+"_location"] = df.at[index_label_inception,'location']
+                            df.drop([index_label_inception], inplace=True)
```

Comment:
I think that... while iterating a list for every item in that list, dropping the occasional item needlessly complicates the code.  Have you considered instead either adding to an empty list called `found_children` or maintaining two lists, one to iterate and one to drop from?

This entire loop (lines 282-389) should probably be rewritten with a different approach.